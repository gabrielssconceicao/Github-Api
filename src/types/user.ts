export interface IUser {
  name: null | string;
  login: string;
  avatar_url: string;
  followers: number;
  following: number;
  location: null | string;
  html_url: string;
  bio: null | string;
  public_repos: number;
  repos_url: string;
}

export interface IUserResponse {
  user: IUser;
  status: number;
}
