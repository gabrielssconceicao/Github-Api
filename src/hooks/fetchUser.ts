import type { IUserResponse, IUser } from '@/types/user';
import { useLoadStore } from '@/stores/load';
export const fetchUser = async (userLogin: string): Promise<IUserResponse> => {
  const loadStore = useLoadStore();
  loadStore.setLoading();
  try {
    const response = await fetch(`https://api.github.com/users/${userLogin}`);
    const data = await response.json();
    return {
      status: response.status,
      user: data,
    };
  } catch (error: any) {
    return {
      status: error.status,
      user: {} as IUser,
    };
  } finally {
    loadStore.setNotLoading();
  }
};
