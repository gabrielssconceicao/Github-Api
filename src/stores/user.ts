import { ref } from 'vue';
import { defineStore } from 'pinia';
import type { IUserResponse } from '@/types/user';
import { fetchUser } from '@/hooks/fetchUser';
export const useUserStore = defineStore('counter', () => {
  const githubUser = ref<IUserResponse>({} as IUserResponse);

  async function searchUser(userLogin: string) {
    const data = await fetchUser(userLogin);
    githubUser.value = data;
  }

  return { githubUser, searchUser };
});
