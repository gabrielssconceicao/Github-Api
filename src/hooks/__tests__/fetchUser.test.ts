import { describe, it, expect } from 'vitest';
import { fetchUser } from '../fetchUser';
import { userResponseError, userResponseSuccess } from '@/mock/__tests__/user';
describe('fetchUser()', () => {
  it('should fetch user data successively', async () => {
    const response = await fetchUser('resolveTest');
    expect(response.status).toEqual(userResponseSuccess.status);
    expect(response.user).toEqual(userResponseSuccess.user);
  });

  it('should fetch user data error', async () => {
    const response = await fetchUser('rejectTest');
    expect(response.status).toEqual(userResponseError.status);
    expect(response.user).toEqual(userResponseError.user);
  });
});
