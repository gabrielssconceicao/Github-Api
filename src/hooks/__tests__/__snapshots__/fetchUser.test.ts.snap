// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchUser() > should fetch user data error 1`] = `
{
  "status": 404,
  "user": {},
}
`;

exports[`fetchUser() > should fetch user data successively 1`] = `
{
  "status": 200,
  "user": {
    "avatar_url": "https://avatars.githubusercontent.com/u/10590076?v=4",
    "bio": "lorem ipsum dolor sit amet",
    "followers": 0,
    "following": 5,
    "html_url": "https://github.com/testeuser",
    "location": "Brazil",
    "login": "testeuser",
    "name": "Teste User",
    "public_repos": 5,
    "repos_url": "https://api.github.com/users/testeuser/repos",
  },
}
`;
